### PR TITLE
Fast add stuff

### PR DESCRIPTION
**Pull request by whyrusleeping**
This is based on the flushing disable PR i put up the other day since I needed some changes from there.

This makes ipfs add use the mfs code under the hood to make adds a bit faster.

### CODE REVIEW

File: test/sharness/t0042-add-skip.sh [hunk 1]
```diff
@@ -38,11 +38,11 @@ test_add_skip() {
 		cat >expected <<-\EOF &&
 			added QmcAREBcjgnUpKfyFmUGnfajA1NQS5ydqRp7WfqZ6JF8Dx planets/.asteroids/ceres.txt
 			added QmZ5eaLybJ5GUZBNwy24AA9EEDTDpA4B8qXnuN3cGxu2uF planets/.asteroids/pallas.txt
-			added Qmf6rbs5GF85anDuoxpSAdtuZPM9D2Yt3HngzjUVSQ7kDV planets/.asteroids
 			added QmaowqjedBkUrMUXgzt9c2ZnAJncM9jpJtkFfgdFstGr5a planets/.charon.txt
 			added QmU4zFD5eJtRBsWC63AvpozM9Atiadg9kPVTuTrnCYJiNF planets/.pluto.txt
 			added QmZy3khu7qf696i5HtkgL2NotsCZ8wzvNZJ1eUdA5n8KaV planets/mars.txt
 			added QmQnv4m3Q5512zgVtpbJ9z85osQrzZzGRn934AGh6iVEXz planets/venus.txt
+			added Qmf6rbs5GF85anDuoxpSAdtuZPM9D2Yt3HngzjUVSQ7kDV planets/.asteroids
```
Comment (jbenet) [inline test/sharness/t0042-add-skip.sh]: we should sort the files, otherwise non-determinism.

Comment (whyrusleeping) [inline test/sharness/t0042-add-skip.sh]: they are sorted on input, but yeah. probably makes sense to sort it in the tests

**Comment by whyrusleeping** [comment]
@jbenet could use some CR here. I'll squash and rework the commits while addressing CR

File: test/sharness/t0042-add-skip.sh (hunk 1 above)
Comment (jbenet) [inline test/sharness/t0042-add-skip.sh]: If the outputs are based on concurrent behavior then I guess we have to give up sorting in the output and that's probably ok.

Comment (whyrusleeping) [inline test/sharness/t0042-add-skip.sh]: right, but theres no concurrency here at this point. Its pretty fast without it

File: commands/cli/parse.go [hunk 2]
```diff
@@ -356,44 +376,34 @@ func appendStdinAsString(args []string, stdin *os.File) ([]string, *os.File, err
 	return append(args, strings.Split(input, "\n")...), nil, nil
 }
 
-func appendFile(args []files.File, inputs []string, argDef *cmds.Argument, recursive bool) ([]files.File, []string, error) {
-	fpath := inputs[0]
+const notRecursiveFmtStr = "'%s' is a directory, use the '-%s' flag to specify directories"
+const dirNotSupportedFmtStr = "Invalid path '%s', argument '%s' does not support directories"
+
+func appendFile(fpath string, argDef *cmds.Argument, recursive bool) (files.File, error) {
```
Comment (jbenet) [inline commands/cli/parse.go]: :clap: thank you very much for the simplification

File: core/commands/add.go [hunk 3]
```diff
@@ -44,6 +45,7 @@ remains to be implemented.
 	Options: []cmds.Option{
 		cmds.OptionRecursivePath, // a builtin option that allows recursive paths (-r, --recursive)
 		cmds.BoolOption(quietOptionName, "q", "Write minimal output"),
+		cmds.BoolOption(silentOptionName, "x", "Write no output"),
```
Comment (jbenet) [inline core/commands/add.go]: i like this. should we do something like `-qq` (kind of like the typical unixy `-v, -vv, -vvv` crap.). would be nice to have the same behavior everywhere:

```
[none]    normal output
-q        minimal ouput
-qq       no output
```
- **argument against `-qq`** - it sounds kind of dumb. (have always thought `-vvvv` "verbose * 4" is very silly)
- **argument against `-{x, s, }`** - would be nice to have the same convention everywhere without burning another character in every single command. `-x` and `-s` are likely to make sense in various places.
- **argument against shorthand** - maybe we dont use `-qq` or `-{s, x}`, but force a full `--silent` for no output.

Comment (whyrusleeping) [inline core/commands/add.go]: i did `qq` first, but apparently it doesnt work for our commands lib. it parses it as just `q` :(

File: core/commands/add.go [hunk 4]
```diff
@@ -65,14 +73,19 @@ remains to be implemented.
 			return nil
```
Comment (jbenet) [inline core/commands/add.go]: we should:

```go
log.Error("progress bar error: %s, err)
```

File: core/commands/add.go [hunk 5]
```diff
@@ -65,14 +73,19 @@ remains to be implemented.
...
-			// see comment above
-			return nil
-		}
+		sizeCh := make(chan int64, 1)
+		req.Values()["size"] = sizeCh
+
+		go func() {
+			size, err := sizeFile.Size()
+			if err != nil {
+				// see comment above
```
Comment (jbenet) [inline core/commands/add.go]: we should:

```go
log.Error("progress bar error: %s, err)
```

File: core/commands/add.go [hunk 6]
```diff
@@ -189,17 +207,23 @@ remains to be implemented.
...
+		if err != nil {
+			res.SetError(u.ErrCast(), cmds.ErrNormal)
+			return
+		}
+
+		var showProgressBar bool
+		if prgFound {
+			showProgressBar = progress
+		} else if !quiet {
+			showProgressBar = true
```
Comment (jbenet) [inline core/commands/add.go]: needs silent consideration, too. how about:
```go
showProgressBar := true
if quiet || silent {
   showProgressBar = false
} else if prgFound {
  showProgressBar = progress
}
```

File: core/commands/add.go [hunk 7]
```diff
@@ -189,17 +207,23 @@ remains to be implemented.
...
 		}
-		showProgressBar := !quiet && size >= progressBarMinSize
 
 		var bar *pb.ProgressBar
 		var terminalWidth int
 		if showProgressBar {
-			bar = pb.New64(size).SetUnits(pb.U_BYTES)
+			bar = pb.New64(0).SetUnits(pb.U_BYTES)
```
Comment (jbenet) [inline core/commands/add.go]: wait, what does that `0` do? doesn't it set the max size?

File: core/commands/add.go [hunk 8]
```diff
@@ -215,43 +239,63 @@ remains to be implemented.
...
 				}
-				if output.Name != lastFile || output.Bytes < lastBytes {
-					prevFiles += lastBytes
-					lastFile = output.Name
+			case size := <-sizeChan:
+				if showProgressBar {
+					bar.Total = size
+					bar.ShowPercent = true
+					bar.ShowBar = true
+					bar.ShowTimeLeft = true
```
Comment (jbenet) [inline core/commands/add.go]: ohhh nice

File: core/commands/add.go (hunk 7 above)
Comment (jbenet) [inline core/commands/add.go]: nvm -- it's updated below.

File: core/coreunix/add.go [hunk 9]
```diff
@@ -153,8 +161,55 @@ func (params *Adder) PinRoot() error {
...
 
-func (params *Adder) Finalize(DAG dag.DAGService) (*dag.Node, error) {
-	return params.editor.Finalize(DAG)
+func (params *Adder) Finalize() (*dag.Node, error) {
+	root, err := params.mr.GetValue().GetNode()
+	if err != nil {
+		return nil, err
+	}
+
+	params.RootNode()
```
Comment (jbenet) [inline core/coreunix/add.go]: why is this call here? since you dont capture the value it looks like it's causing some sideeffect, maybe comment?

File: core/coreunix/add.go [hunk 10]
```diff
@@ -241,19 +318,22 @@ func (params *Adder) addNode(node *dag.Node, path string) error {
...
-	return outputDagnode(params.out, path, node)
+	if !params.Silent {
+		return outputDagnode(params.out, path, node)
+	}
+	return nil
 }
 
 // Add the given file while respecting the params.
-func (params *Adder) AddFile(file files.File) (*dag.Node, error) {
+func (params *Adder) AddFile(file files.File) error {
```
Comment (jbenet) [inline core/coreunix/add.go]: why no longer return dagnode? not useful anymore?

Comment (jbenet) [inline core/coreunix/add.go]: what if we only add one file? with no wrap? (i guess there is a dir and we unwrap it)

File: core/coreunix/add.go [hunk 11]
```diff
@@ -379,7 +446,7 @@ func getOutput(dagnode *dag.Node) (*Object, error) {
 	for i, link := range dagnode.Links {
 		output.Links[i] = Link{
 			Name: link.Name,
-			Hash: link.Hash.B58String(),
+			//Hash: link.Hash.B58String(),
```
Comment (jbenet) [inline core/coreunix/add.go]: hah, so much silly b58.

File: core/coreunix/add_test.go [hunk 12]
```diff
@@ -29,9 +23,9 @@ func TestAddRecursive(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	if k, err := AddR(node, path.Join(here, "test_data")); err != nil {
+	if k, err := AddR(node, "test_data"); err != nil {
```
Comment (jbenet) [inline core/coreunix/add_test.go]: this implies something changed in the logic of `AddR` -- what was that?

Comment (jbenet) [inline core/coreunix/add_test.go]: (i didn't notice after reading it :/ )

File: exchange/bitswap/workers.go [hunk 13]
```diff
@@ -89,7 +89,7 @@ func (bs *Bitswap) provideWorker(px process.Process) {
 		defer cancel()
 
 		if err := bs.network.Provide(ctx, k); err != nil {
-			log.Error(err)
+			//log.Error(err)
```
Comment (jbenet) [inline exchange/bitswap/workers.go]: not sure this shouldnt be logged at all. are you trying to silence a particular error?

File: core/coreunix/add_test.go (hunk 12 above)
Comment (whyrusleeping) [inline core/coreunix/add_test.go]: yeah, what changed was how the paths are consumed. I think this is erroneous, I'll fix it before we merge this PR.

File: exchange/bitswap/workers.go (hunk 13 above)
Comment (whyrusleeping) [inline exchange/bitswap/workers.go]: whoops, temp code i forgot to remove. Was running with no bootstrap peers and bitswap telling me that constantly was rather annoying.

File: test/sharness/t0042-add-skip.sh (hunk 1 above)
Comment (jbenet) [inline test/sharness/t0042-add-skip.sh]: yeah let's sort on tests. we should probably warn in `ipfs add --help` that the output order is not deterministic.

File: test/sharness/t0080-repo.sh [hunk 14]
```diff
@@ -29,11 +29,6 @@ test_expect_success "'ipfs repo gc' succeeds" '
 	ipfs repo gc >gc_out_actual
 '
 
-test_expect_success "'ipfs repo gc' looks good (patch root)" '
-	PATCH_ROOT=QmQXirSbubiySKnqaFyfs5YzziXRB5JEVQVjU6xsd7innr &&
-	grep "removed $PATCH_ROOT" gc_out_actual
```
Comment (jbenet) [inline test/sharness/t0080-repo.sh]: why is this removed? not understanding why it no longer removes this root node

File: test/sharness/t0250-files-api.sh [hunk 15]
```diff
@@ -316,13 +316,26 @@ test_files_api() {
 		verify_dir_contents /cats file1 ipfs this
 	'
 
+	test_expect_success "write 'no-flush' succeeds" '
+		echo "testing" | ipfs files write -f -e /cats/walrus
+	'
+
+	test_expect_success "changes bubbled up to root on inspection" '
+		ipfs files stat / | head -n1 > root_hash
```
Comment (jbenet) [inline test/sharness/t0250-files-api.sh]: can we test that they _didn't_ bubble up beforehand? are there ops that will read the unsynced thing? also-- it makes me uneasy that sync requires another op hit, no matter how long time passes. i would imagine that `--sync=false` is an optimization for heavy load, but that a `sync` would happen if time passed. (like an os buffer cache)

Comment (jbenet) [inline test/sharness/t0250-files-api.sh]: actually, can we test: 
- that they didn't bubble up beforehand?
- that the dag objects themselves are written and pinned (for safety)

**Comment by jbenet** [comment]
@whyrusleeping some comments above -- otherwise this all LGTM. fantastic improvement!

File: test/sharness/t0250-files-api.sh (hunk 15 above)
Comment (whyrusleeping) [inline test/sharness/t0250-files-api.sh]: i guess i can do it manually and figure out what the bubbled up hash would be, then check to make sure it doesnt exist.

Comment (jbenet) [inline test/sharness/t0250-files-api.sh]: :+1:

File: core/commands/add.go (hunk 3 above)
Comment (jbenet) [inline core/commands/add.go]: ok with just doing `--silent` for now?

File: core/coreunix/add.go [hunk 16]
```diff
@@ -241,96 +329,111 @@ func (params *Adder) addNode(node *dag.Node, path string) error {
...
 		if err != nil && err != io.EOF {
-			return nil, err
+			return err
 		}
 		if file == nil {
 			break
 		}
 
-		node, err := params.AddFile(file)
+		err = params.addFile(file)
```
Comment (jbenet) [inline core/coreunix/add.go]: ok so i think directories should add every file in a new goroutine (with rate limiting). i had started on this earlier, but aborted. the general idea was every dir -> files fan out uses goroutines, like:

```go
ctx, cancel := context.WithCancel(params.ctx)
defer cancel()
// this context will cancel out all the children if we error out early.
// would need to wire the addFile funcs with a ctx. OR we _could_ cancel 
// the entire context of thea adder instead...

errs := make(chan error, 10) // some room to avoid waiting for errors
                // dont think we have access to read len(dir) though

for {
  file, err := dir.NextFile()
  if err != nil && err != io.EOF {
    return err // defer cancel unblocks other children
  }
  if file == nil {
    break // done with files
  }

  <-params.concurrency // adder-global parameter
  go func(file) {
    defer func() {
      params.concurrency<- struct{}{} // done
    }()

    err := params.addFile(ctx, file) // ctx cancel should unblock addFile

    // return even nil errors, so we force the parent to wait
    // until _all_ children are done during non-error ops.
    // but bail on ctx.Done.
    select {
    case errs<- err:
    case ctx.Done():
    }
  }(file)
}

for err := range errs {
  if err == nil {
    continue
  }
  if _, ok := err.(*hiddenFileError); ok {
    continue // hidden file error, skip file
  }
  // error!
  return err // defer cancel unblocks + terminates all children.
}

// all children done.
return nil
```

**Comment by jbenet** [comment]
@whyrusleeping general comment: i think adding files should be concurrent because a single add should leverage as much concurrent io + hashing as possible (right now it _sequentially_ trades off between io and hashing).

I think adding https://github.com/ipfs/go-ipfs/pull/2039/files#r46767388 will fix it

File: core/coreunix/add.go (hunk 16 above)
Comment (whyrusleeping) [inline core/coreunix/add.go]: this is hard to do right, this code will deadlock when it encounters a directory deeper than your concurrency factor.

Comment (whyrusleeping) [inline core/coreunix/add.go]: we could add something like:
```
if file.IsDir() {
    params.addFile(ctx, file)
} else {
   // concurrency thing
}

Comment (jbenet) [inline core/coreunix/add.go]: yeah i realized that after posting, we can fix that by bumping up the concurrency level one notch in dirs (i.e. semaup at the top of inside `addDir`, and defer semadown) -- better than this one o/ because this doesn't allow concurrent add of dirs-- and complicates the code by doing two different things (double the error handling, etc).

File: test/sharness/t0250-files-api.sh (hunk 15 above)
Comment (whyrusleeping) [inline test/sharness/t0250-files-api.sh]: one thing to note: the `flush=false` option does nothing when running without a daemon. So i'm going to have to special case the test

**Comment by whyrusleeping** [comment]
@jbenet would you be opposed to having the concurrency stuff done in a separate PR? This one is already getting pretty large

**Comment by jbenet** [comment]
@whyrusleeping my thought is "it's almost there!" -- but sure, if it's easier for you

File: test/sharness/t0043-add-w.sh [hunk 17]
```diff
@@ -27,20 +27,20 @@ added QmYC3u5jGWuyFwvTxtvLYm2K3SpWZ31tg3NjpVVvh9cJaJ _jo7/wzvsihy
...
-added QmVPwNy8pZegpsNmsjjZvdTQn4uCeuZgtzhgWhRSQWjK9x gnz66h
 added QmPcaX84tDiTfzdTn8GQxexodgeWH6mHjSss5Zfr5ojssb _jo7/-s782qgs
 added QmaVBqquUuXKjkyWHXaXfsaQUxAnsCKS95VRDHU8PzGA4K _jo7/15totauzkak-
 added QmaAHFG8cmhW3WLjofx5siSp44VV25ETN6ThzrU8iAqpkR _jo7/galecuirrj4r
 added QmeuSfhJNKwBESp1W9H8cfoMdBfW3AeHQDWXbNXQJYWp53 _jo7/mzo50r-1xidf5zx
 added QmYC3u5jGWuyFwvTxtvLYm2K3SpWZ31tg3NjpVVvh9cJaJ _jo7/wzvsihy
+added QmVaKAt2eVftNKFfKhiBV7Mu5HjCugffuLqWqobSSFgiA7 h3qpecj0
 added QmQkib3f9XNX5sj6WEahLUPFpheTcwSRJwUCSvjcv8b9by _jo7
-added Qme987pqNBhZZXy4ckeXiR7zaRQwBabB7fTgHurW2yJfNu 4r93
+added QmVPwNy8pZegpsNmsjjZvdTQn4uCeuZgtzhgWhRSQWjK9x gnz66h
```
Comment (jbenet) [inline test/sharness/t0043-add-w.sh]: still needs sorting?

Comment (jbenet) [inline test/sharness/t0043-add-w.sh]: or, i guess it doesn't matter what order they're in? maybe just restore the original order to make sure/

File: test/sharness/t0045-ls.sh [hunk 18]
```diff
@@ -27,12 +27,12 @@ test_ls_cmd() {
...
 			added QmQNd6ubRXaNG6Prov8o6vk3bn6eWsj9FxLGrAVDUAGkGe testData/d1/128
 			added QmZULkCELmmk5XNfCgTnCyFgAVxBRBXyDHGGMVoLFLiXEN testData/d1/a
-			added QmSix55yz8CzWXf5ZVM9vgEvijnEeeXiTSarVtsqiiCJss testData/d1
 			added QmbQBUSRL9raZtNXfpTDeaxQapibJEG6qEY8WqAN22aUzd testData/d2/1024
 			added QmaRGe7bVmVaLmxbrMiVNXqW4pRNNp3xq7hFtyRKA3mtJL testData/d2/a
-			added QmR3jhV4XpxxPjPT3Y8vNnWvWNvakdcT3H6vqpRBsX1MLy testData/d2
 			added QmeomffUNfmQy76CQGy9NdmqEnnHU9soCexBnGU3ezPHVH testData/f1
 			added QmNtocSs7MoDkJMc1RkyisCSKvLadujPsfJfSdJ3e1eA1M testData/f2
+			added QmSix55yz8CzWXf5ZVM9vgEvijnEeeXiTSarVtsqiiCJss testData/d1
+			added QmR3jhV4XpxxPjPT3Y8vNnWvWNvakdcT3H6vqpRBsX1MLy testData/d2
```
Comment (jbenet) [inline test/sharness/t0045-ls.sh]: keep orignal order + test sorted?

File: test/sharness/t0200-unixfs-ls.sh [hunk 19]
```diff
@@ -27,12 +27,12 @@ test_ls_cmd() {
...
 			added QmQNd6ubRXaNG6Prov8o6vk3bn6eWsj9FxLGrAVDUAGkGe testData/d1/128
 			added QmZULkCELmmk5XNfCgTnCyFgAVxBRBXyDHGGMVoLFLiXEN testData/d1/a
-			added QmSix55yz8CzWXf5ZVM9vgEvijnEeeXiTSarVtsqiiCJss testData/d1
 			added QmbQBUSRL9raZtNXfpTDeaxQapibJEG6qEY8WqAN22aUzd testData/d2/1024
 			added QmaRGe7bVmVaLmxbrMiVNXqW4pRNNp3xq7hFtyRKA3mtJL testData/d2/a
-			added QmR3jhV4XpxxPjPT3Y8vNnWvWNvakdcT3H6vqpRBsX1MLy testData/d2
 			added QmeomffUNfmQy76CQGy9NdmqEnnHU9soCexBnGU3ezPHVH testData/f1
 			added QmNtocSs7MoDkJMc1RkyisCSKvLadujPsfJfSdJ3e1eA1M testData/f2
+			added QmSix55yz8CzWXf5ZVM9vgEvijnEeeXiTSarVtsqiiCJss testData/d1
+			added QmR3jhV4XpxxPjPT3Y8vNnWvWNvakdcT3H6vqpRBsX1MLy testData/d2
```
Comment (jbenet) [inline test/sharness/t0200-unixfs-ls.sh]: test sorted?

Comment (whyrusleeping) [inline test/sharness/t0200-unixfs-ls.sh]: the output is actually deterministic. The order just changed

Comment (jbenet) [inline test/sharness/t0200-unixfs-ls.sh]: ahhh ok makes sense now (dirs)

**Comment by whyrusleeping** [comment]
good to merge?

**Comment by jbenet** [comment]
@whyrusleeping Still missing:

- [ ] https://github.com/ipfs/go-ipfs/pull/2039#discussion_r46766019 - https://github.com/ipfs/go-ipfs/pull/2039/files#r46766019
- [ ] https://github.com/ipfs/go-ipfs/pull/2039#discussion_r46766078 https://github.com/ipfs/go-ipfs/pull/2039/files#r46766078
- [ ] https://github.com/ipfs/go-ipfs/pull/2039#discussion_r46766194 https://github.com/ipfs/go-ipfs/pull/2039/files#r46766194

File: core/coreunix/add.go (hunk 9 above)
Comment (whyrusleeping) [inline core/coreunix/add.go]: i think i just forgot to remove that... good catch.  I try not to ever have weird side effects like that.

as an aside: I wish go would let you say "count unhandled error returns as compile errors"

File: core/coreunix/add.go (hunk 10 above)
Comment (whyrusleeping) [inline core/coreunix/add.go]: I changed the role of this function. The `AddFile` call may be made multiple times, when you're done, call `Finalize`.

File: core/commands/add.go (hunk 6 above)
Comment (jbenet) [inline core/commands/add.go]: addressed

**Comment by rht** [comment]
It sure is fast
outdata
(filesize 1KB each)

Memory growth:
memory

However, it is still possible to get to ~O(git) soon if intermediate root nodes are not created (even if the hashings are delayed in this PR) https://github.com/ipfs/go-ipfs/pull/1964#issuecomment-156885556.

**Comment by jbenet** [comment]
great stats, thanks @rht.

would be awesome to get these auto-generated by running one script. could
add it under `tests/bench/` or something
